php/integrations/yoast.php

```diff
@@ -0,0 +1,243 @@
+		$graph_piece_generator->context->canonical      = $author_url;
+		$graph_piece_generator->context->main_schema_id = $author_url;
+
+		return $graph_piece_generator->generate();
+	}
+
+	/**
+	 * Filters the graph output to add authors.
+	 *
+	 * @param array                   $data                   The schema graph.
+	 * @param Meta_Tags_Context       $context                The context object.
+	 * @param Abstract_Schema_Piece   $graph_piece_generator  The graph piece generator.
+	 * @param Abstract_Schema_Piece[] $graph_piece_generators The graph piece generators.
+	 *
+	 * @return array The (potentially altered) schema graph.
+	 */
+	public static function filter_author_graph( $data, $context, $graph_piece_generator, $graph_piece_generators ) {
+		if ( ! isset( $data['image']['url'] ) ) {
+			return $data;
+		}
+
+		if ( isset( $data['image']['@id'] ) ) {
+			$data['image']['@id'] .= \md5( $data['image']['url'] );
+		}
+
+		if ( isset( $data['logo']['@id'] ) ) {
+			$data['logo']['@id'] .= \md5( $data['image']['url'] );
+		}
+
+		return $data;
+	}
+
+	/**
+	 * Filters the graph output to add authors.
+	 *
+	 * @param array             $data    The schema graph.
+	 * @param Meta_Tags_Context $context Context object.
+	 *
+	 * @return array The (potentially altered) schema graph.
+	 */
+	public static function filter_graph( $data, $context ) {
+		if ( ! \is_singular() ) {
+			return $data;
+		}
+
+		if ( ! \function_exists( '\get_coauthors' ) ) {
+			return $data;
+		}
+
+		/**
+		 * Contains the authors from the CoAuthors Plus plugin.
+		 *
+		 * @var \WP_User[] $author_objects
+		 */
+		$author_objects = \get_coauthors( $context->post->ID );
+
+		$ids     = [];
+		$authors = [];
+
+		// Add the authors to the schema.
+		foreach ( $author_objects as $author ) {
+			$author_generator          = new CoAuthor();
+			$author_generator->context = $context;
+			$author_generator->helpers = \YoastSEO()->helpers;
+
+			if ( $author instanceof \WP_User ) {
+				$author_data = $author_generator->generate_from_user_id( $author->ID );
+			} elseif ( ! empty( $author->type ) && $author->type === 'guest-author' ) {
+				$author_data = $author_generator->generate_from_guest_author( $author );
+			}
+
+			if ( ! empty( $author_data ) ) {
+				$ids[]     = [ '@id' => $author_data['@id'] ];
+				$authors[] = $author_data;
+			}
+		}
+		$schema_types  = new Schema_Types();
+		$article_types = $schema_types->get_article_type_options_values();
+
+		// Change the author reference to reference our multiple authors.
+		$add_to_graph = false;
+		foreach ( $data as $key => $piece ) {
+			if ( \in_array( $piece['@type'], $article_types, true ) ) {
+				$data[ $key ]['author'] = $ids;
+				$add_to_graph           = true;
+				break;
+			}
+		}
+
+		if ( $add_to_graph ) {
+			// Clean all Persons from the schema, as the user stored as post owner might be incorrectly added if the post post has only guest authors as authors.
+			$data = \array_filter(
```

Comment:
Very much a nitpick, but our PHP convention in Newspack is to not globally namespace core PHP functions—only global functions like those from WP or other layers on top of PHP.

Reply:
This comes by influence of how they do things in Yoast, and since I was writing it for Yoast I tried to follow their standards. I'm gonna check and see how we do this in CAP

Reply:
Cleaned up some namespacing of global functions in https://github.com/Automattic/Co-Authors-Plus/pull/877/commits/ff9db8206c398d7421362ff6c7e1a86e21e155d5

Reply:
Nice, I see that CAP does not use the global namespace, so this looks good to me.